Creating explicit array from return values gives c error: declaration of an array of incomplete type elements

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
V full version: V 0.4.8 5c65e58.561efd1
OS: windows, Microsoft Windows 11 Pro v22631 64-bit
Processor: 24 cpus, 64bit, little endian, 

getwd: D:\vbug\src
vexe: C:\v\v.exe
vexe mtime: 2024-11-15 16:27:23

vroot: OK, value: C:\v
VMODULES: OK, value: C:\Users\info\.vmodules
VTMP: OK, value: C:\Users\info\Documents\MobaXterm\slash\mx86_64b\var\log\xwin\v_0

Git version: git version 2.42.0.windows.2
Git vroot status: weekly.2024.46-29-g561efd14-dirty
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,

operable program or batch file.


emcc version: Error: 'emcc' is not recognized as an internal or external command,

operable program or batch file.


thirdparty/tcc status: thirdparty-windows-amd64 b425ac82
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg main.v`

```v
module main

import math.vec

type Vec4 = vec.Vec4[f32]

fn (v Vec4) to_array() [4]f32 {
	return [v.x, v.y, v.z, v.w]!
}

struct Uniforms {
	lights [2][4]f32
}

fn main() {
	v := Vec4{0, 0, 0, 1}
	uniforms := Uniforms{
		lights: [
			v.to_array(),
			v.to_array(),
		]!
	}
	eprintln('uniforms: ${uniforms}')
}
```

**What did you expect to see?**

no c error

**What did you see instead?**

```
================== C compilation error (from tcc): ==============
cc: C:/Users/info/Documents/MobaXterm/slash/mx86_64b/var/log/xwin/v_0/main.01JCRAM1S9SZP5KHX88TYAX6TN.tmp.c:540: warning: WINVER redefined
cc: C:/Users/info/Documents/MobaXterm/slash/mx86_64b/var/log/xwin/v_0/main.01JCRAM1S9SZP5KHX88TYAX6TN.tmp.c:1354: error: declaration of an array of incomplete type elements
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzM3NzdlZmU3Y2M1YTc4NTQyZDI4ZWUiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.3TgHxLjcv8cXNVbXF9uPfFquVrpqQr22KL0bDQdl3yA">Huly&reg;: <b>V_0.6-21309</b></a></sub>

Even this small section alone fails:
https://play.vlang.io/p/48e7b512e0
```v
import math.vec

type Vec4 = vec.Vec4[f32]
```
```
Running code...
Can't run code. The server returned an error:
/tmp/v_60000/code.01JCRC2FREFXQE8GR1700MCVYM.tmp.c:2223: error: invalid type
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
Exited with error status 1
Please try again.

```
 or maybe is still another issue

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2561,20 +2561,26 @@ fn (mut g Gen) expr_with_var(expr ast.Expr, got_type_raw ast.Type, expected_type
 	styp := g.styp(expected_type)
 
 	g.writeln('${styp} ${tmp_var};')
-	if expr is ast.ArrayInit && expr.is_fixed && expr.exprs.len > 0
-		&& expr.exprs.any(it is ast.CallExpr) {
-		// [ foo(), foo() ]!
-		val_typ := g.table.value_type(got_type_raw)
-		for i, item_expr in expr.exprs {
-			g.write('memcpy(${tmp_var}[${i}], ')
-			g.expr(item_expr)
-			g.writeln(', sizeof(${g.styp(val_typ)}));')
-		}
-	} else {
-		// regular init
-		g.write('memcpy(&${tmp_var}, ')
-		g.expr(expr)
-		g.writeln(', sizeof(${styp}));')
+	g.write('memcpy(&${tmp_var}, ')
+	g.expr(expr)
+	g.writeln(', sizeof(${styp}));')
+	g.write(stmt_str)
+	return tmp_var
+}
+
+// expr_with_fixed_array generates code for fixed array initialization with expr which requires tmp var
+fn (mut g Gen) expr_with_fixed_array(expr ast.ArrayInit, got_type_raw ast.Type, expected_type ast.Type) string {
+	stmt_str := g.go_before_last_stmt().trim_space()
+	g.empty_line = true
+	tmp_var := g.new_tmp_var()
+	styp := g.styp(expected_type)
+	g.writeln('${styp} ${tmp_var};')
+	// [ foo(), foo() ]!
+	val_typ := g.table.value_type(got_type_raw)
+	for i, item_expr in expr.exprs {
+		g.write('memcpy(${tmp_var}[${i}], ')
+		g.expr(item_expr)
+		g.writeln(', sizeof(${g.styp(val_typ)}));')
 	}
 	g.write(stmt_str)
 	return tmp_var

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2559,10 +2559,23 @@ fn (mut g Gen) expr_with_var(expr ast.Expr, got_type_raw ast.Type, expected_type
 	g.empty_line = true
 	tmp_var := g.new_tmp_var()
 	styp := g.styp(expected_type)
+
 	g.writeln('${styp} ${tmp_var};')
-	g.write('memcpy(&${tmp_var}, ')
-	g.expr(expr)
-	g.writeln(', sizeof(${styp}));')
+	if expr is ast.ArrayInit && expr.is_fixed && expr.exprs.len > 0
+		&& expr.exprs.any(it is ast.CallExpr) {
+		// [ foo(), foo() ]!
+		val_typ := g.table.value_type(got_type_raw)
+		for i, item_expr in expr.exprs {
+			g.write('memcpy(${tmp_var}[${i}], ')
+			g.expr(item_expr)
+			g.writeln(', sizeof(${g.styp(val_typ)}));')
+		}
+	} else {
+		// regular init
+		g.write('memcpy(&${tmp_var}, ')
+		g.expr(expr)
+		g.writeln(', sizeof(${styp}));')
+	}
 	g.write(stmt_str)
 	return tmp_var
 }

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2620,7 +2620,7 @@ fn (mut g Gen) expr_with_cast(expr ast.Expr, got_type_raw ast.Type, expected_typ
 				// Do not allocate for `Interface(unsafe{nil})` casts
 				is_nil_cast := expr is ast.UnsafeExpr && expr.expr is ast.Nil
 				if is_nil_cast {
-					g.write2('/*nili*/', '((void*)0)')
+					g.write2('/*nil*/', '((void*)0)')
 					return
 				}
 			}
@@ -6901,7 +6901,7 @@ fn (mut g Gen) write_types(symbols []&ast.TypeSymbol) {
 											base)};')
 									}
 								}
-							} else {
+							} else if !(elem_sym.info is ast.ArrayFixed && elem_sym.info.is_fn_ret) {
 								g.type_definitions.writeln('typedef ${fixed_elem_name} ${styp} [${len}];')
 							}
 						}

--- a/vlib/v/gen/c/struct.v
+++ b/vlib/v/gen/c/struct.v
@@ -679,33 +679,54 @@ fn (mut g Gen) struct_init_field(sfield ast.StructInitField, language ast.Langua
 
 		field_unwrap_typ := g.unwrap_generic(sfield.typ)
 		field_unwrap_sym := g.table.final_sym(field_unwrap_typ)
-		if field_unwrap_sym.kind == .array_fixed && sfield.expr in [ast.Ident, ast.SelectorExpr] {
-			info := field_unwrap_sym.info as ast.ArrayFixed
-			g.fixed_array_var_init(g.expr_string(sfield.expr), sfield.expr.is_auto_deref_var(),
-				info.elem_type, info.size)
-		} else if field_unwrap_sym.kind == .array_fixed && (sfield.expr is ast.CallExpr
-			|| (sfield.expr is ast.ArrayInit && sfield.expr.has_index)) {
-			info := field_unwrap_sym.info as ast.ArrayFixed
-			tmp_var := g.expr_with_var(sfield.expr, sfield.typ, sfield.expected_type)
-			g.fixed_array_var_init(tmp_var, false, info.elem_type, info.size)
+		is_auto_deref_var := sfield.expr.is_auto_deref_var()
+		if field_unwrap_sym.info is ast.ArrayFixed {
+			match sfield.expr {
+				ast.Ident, ast.SelectorExpr {
+					g.fixed_array_var_init(g.expr_string(sfield.expr), is_auto_deref_var,
+						field_unwrap_sym.info.elem_type, field_unwrap_sym.info.size)
+				}
+				ast.CallExpr {
+					tmp_var := g.expr_with_var(sfield.expr, sfield.typ, sfield.expected_type)
+					g.fixed_array_var_init(tmp_var, false, field_unwrap_sym.info.elem_type,
+						field_unwrap_sym.info.size)
+				}
+				ast.ArrayInit {
+					if sfield.expr.has_index
+						|| (sfield.expr.exprs.len > 0 && sfield.expr.exprs.any(it is ast.CallExpr)) {
+						tmp_var := g.expr_with_var(sfield.expr, sfield.typ, sfield.expected_type)
+						g.fixed_array_var_init(tmp_var, false, field_unwrap_sym.info.elem_type,
+							field_unwrap_sym.info.size)
+					} else {
+						g.struct_init_field_default(field_unwrap_typ, sfield)
+					}
+				}
+				else {
+					g.struct_init_field_default(field_unwrap_typ, sfield)
+				}
+			}
 		} else {
-			if field_unwrap_typ != ast.voidptr_type && field_unwrap_typ != ast.nil_type
-				&& (sfield.expected_type.is_ptr() && !sfield.expected_type.has_flag(.shared_f))
-				&& !sfield.expected_type.has_flag(.option)
-				&& !field_unwrap_typ.is_any_kind_of_pointer() && !field_unwrap_typ.is_number() {
-				g.write('/* autoref */&')
-			}
-
-			if (sfield.expected_type.has_flag(.option) && !field_unwrap_typ.has_flag(.option))
-				|| (sfield.expected_type.has_flag(.result) && !field_unwrap_typ.has_flag(.result)) {
-				g.expr_with_opt(sfield.expr, field_unwrap_typ, sfield.expected_type)
-			} else if sfield.expr is ast.LambdaExpr && sfield.expected_type.has_flag(.option) {
-				g.expr_opt_with_cast(sfield.expr, field_unwrap_typ, sfield.expected_type)
-			} else {
-				g.left_is_opt = true
-				g.expr_with_cast(sfield.expr, field_unwrap_typ, sfield.expected_type)
-			}
+			g.struct_init_field_default(field_unwrap_typ, sfield)
 		}
 		g.inside_cast_in_heap = inside_cast_in_heap // restore value for further struct inits
+	}
+}
+
+fn (mut g Gen) struct_init_field_default(field_unwrap_typ ast.Type, sfield &ast.StructInitField) {
+	if field_unwrap_typ != ast.voidptr_type && field_unwrap_typ != ast.nil_type
+		&& (sfield.expected_type.is_ptr() && !sfield.expected_type.has_flag(.shared_f))
+		&& !sfield.expected_type.has_flag(.option) && !field_unwrap_typ.is_any_kind_of_pointer()
+		&& !field_unwrap_typ.is_number() {
+		g.write('/* autoref */&')
+	}
+
+	if (sfield.expected_type.has_flag(.option) && !field_unwrap_typ.has_flag(.option))
+		|| (sfield.expected_type.has_flag(.result) && !field_unwrap_typ.has_flag(.result)) {
+		g.expr_with_opt(sfield.expr, field_unwrap_typ, sfield.expected_type)
+	} else if sfield.expr is ast.LambdaExpr && sfield.expected_type.has_flag(.option) {
+		g.expr_opt_with_cast(sfield.expr, field_unwrap_typ, sfield.expected_type)
+	} else {
+		g.left_is_opt = true
+		g.expr_with_cast(sfield.expr, field_unwrap_typ, sfield.expected_type)
 	}
 }

--- a/vlib/v/gen/c/struct.v
+++ b/vlib/v/gen/c/struct.v
@@ -692,8 +692,11 @@ fn (mut g Gen) struct_init_field(sfield ast.StructInitField, language ast.Langua
 						field_unwrap_sym.info.size)
 				}
 				ast.ArrayInit {
-					if sfield.expr.has_index
-						|| (sfield.expr.exprs.len > 0 && sfield.expr.exprs.any(it is ast.CallExpr)) {
+					if sfield.expr.has_index {
+						tmp_var := g.expr_with_var(sfield.expr, sfield.typ, sfield.expected_type)
+						g.fixed_array_var_init(tmp_var, false, field_unwrap_sym.info.elem_type,
+							field_unwrap_sym.info.size)
+					} else if sfield.expr.exprs.len > 0 && sfield.expr.exprs.any(it is ast.CallExpr) {
 						tmp_var := g.expr_with_fixed_array(sfield.expr, sfield.typ, sfield.expected_type)
 						g.fixed_array_var_init(tmp_var, false, field_unwrap_sym.info.elem_type,
 							field_unwrap_sym.info.size)

--- a/vlib/v/gen/c/struct.v
+++ b/vlib/v/gen/c/struct.v
@@ -694,7 +694,7 @@ fn (mut g Gen) struct_init_field(sfield ast.StructInitField, language ast.Langua
 				ast.ArrayInit {
 					if sfield.expr.has_index
 						|| (sfield.expr.exprs.len > 0 && sfield.expr.exprs.any(it is ast.CallExpr)) {
-						tmp_var := g.expr_with_var(sfield.expr, sfield.typ, sfield.expected_type)
+						tmp_var := g.expr_with_fixed_array(sfield.expr, sfield.typ, sfield.expected_type)
 						g.fixed_array_var_init(tmp_var, false, field_unwrap_sym.info.elem_type,
 							field_unwrap_sym.info.size)
 					} else {

--- a/vlib/v/tests/fixed_array_2_dims_init_test.v
+++ b/vlib/v/tests/fixed_array_2_dims_init_test.v
@@ -5,14 +5,18 @@ struct Uniforms {
 }
 
 fn r() [4]f32 {
-	return [4]f32{}
+	return [f32(1.1), 1.2, 1.3, 1.4]!
 }
 
 fn test_main() {
-	_ := Uniforms{
+	v := Uniforms{
 		lights: [
 			r(),
 			r(),
 		]!
 	}
+	assert v.lights[0][0] == f32(1.1)
+	assert v.lights[0][1] == f32(1.2)
+	assert v.lights[0][2] == f32(1.3)
+	assert v.lights[0][3] == f32(1.4)
 }

--- a/vlib/v/tests/fixed_array_2_dims_init_test.v
+++ b/vlib/v/tests/fixed_array_2_dims_init_test.v
@@ -1,0 +1,18 @@
+module main
+
+struct Uniforms {
+	lights [2][4]f32
+}
+
+fn r() [4]f32 {
+	return [4]f32{}
+}
+
+fn test_main() {
+	_ := Uniforms{
+		lights: [
+			r(),
+			r(),
+		]!
+	}
+}

